Playwright: move types and interfaces into src/types/ directory.

worldomonation opened the pull request:
#### Proposed Changes

It's ya boi @worldomonation back again with a big refactor!

This PR moves as many `type` and `interface` declarations into its own type declaration files nearby to where the types/interfaces were initially declared.

Key changes:
- move most of the types and interfaces into `/types` directory.
- update specs, POMs and unit tests as required.

Miscellaneous changes:
- move two unit tests from `calypso-e2e/tests/libs/utils` to `calypso-e2e/src/libs/utils/tests` to make the type imports much shorter and "closer".

#### Testing Instructions

*Everything* should pass. Literally everything:
  - [x] Gutenberg E2E (desktop)
  - [x] Gutenberg E2E (mobile)
  - [x] Editor Tracking (desktop)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E
  - [x] i18n E2E
  - [x] P2 E2E
  - [x] Unit Tests

Related to #

worldomonation committed:
Move RestAPIClient types and interfaces into <root>/types/rest-api-client.types.ts

worldomonation committed:
Move types from env-variables.ts into its own file.

worldomonation committed:
Move DataHelper and rest of Types into its own files.

worldomonation committed:
Fix failure on Editor Tracing: Block Inserted spec.

worldomonation committed:
Move Secrets types.

worldomonation committed:
Move types and interfaces from components/pages.

worldomonation committed:
Remove the `Site` interface.

worldomonation committed:
Add missing "s" in the file name for secrets-manager-types.ts

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -363,3 +333,39 @@ export function createSuiteTitle( title: string ): string {
+ *
+ * @returns {string[]} Array of strings in ISO-639-1 standard.
+ */
+export function getMag16Locales(): string[] {
+	return [
+		'en',
+		'es',
+		'pt-br',
+		'de',
+		'fr',
+		'he',
+		'ja',
+		'it',
+		'nl',
+		'ru',
+		'tr',
+		'id',
+		'zh-cn',
+		'zh-tw',
+		'ko',
+		'ar',
+		'sv',
+	];
+}
+
+/**
+ * Returns the list of viewports that are available for the framework.
+ *
+ * @returns {string([])} Array of strings for valid viewports.
+ */
+export function getViewports(): string[] {
+	return [ 'mobile', 'desktop' ];
+}
```
I decided to move the "knowledge" of the Mag16 Locales and Viewports out of EnvVars to DataHelper, but as you can see later in the PR I did not follow this convention when dealing with SecretsManager.

To be honest, I am a little uneasy with this as well. I've looked at a few other OSS that use TypeScript (eg. [Jest](https://github.com/facebook/jest/blob/bb0956589f9dbddcc13db0f42b2d4d9ff642b2cd/packages/jest-expect/src/types.ts)) and I don't feel easy about defining and exporting a `const` from the types file. 

At the same time, the way our framework uses the values of a defined array in-code and converted into type means we can't get away from having a `const` defined  _somewhere_.

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -47,12 +20,15 @@ export type CreditCardIssuers = 'Visa';
+}: {
+	mailosaurInbox?: keyof Secrets[ 'mailosaur' ];
+	usernamePrefix?: string;
+} = {} ): NewTestUserDetails {
```
The name here changed resulting in the reformatted code.

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -121,7 +97,7 @@ export function getUsername( { prefix = '' }: { prefix?: string } = {} ): string
  */
 export function getDateString( format: DateFormat ): string | null {
-	if ( format === 'ISO' ) {
+	if ( format === 'ISO-8601' ) {
```
I added the `-8601` postfix to make the format specific.

Not that it really matters; this method is [not used anywhere](https://github.com/Automattic/wp-calypso/search?q=getDateString).

worldomonation commented on packages/calypso-e2e/src/jest-playwright-config/playwright-config.ts:
```diff
@@ -12,6 +12,8 @@ const getTargetDeviceOptions = () => {
 		case 'desktop':
 			deviceName = 'Desktop Chrome HiDPI';
 			break;
+		default:
+			throw new Error( 'No viewport specified.' );
```
Not sure how this didn't throw up compilation error prior to this PR. In any case, moving the `VIEWPORT_NAME` out of `envVariables` started to trigger that a default case is not present.

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -1,8 +1,7 @@
 import { translateFromPage } from '../utils';
-
-export type PreviewOptions = 'Desktop' | 'Mobile' | 'Tablet';
+import type { EditorPreviewOptions } from '../../types/editor-toolbar-component.types';
```
Made the name `PreviewOptions` a bit more specific.

worldomonation committed:
Remove use of `typeof` to define the ViewportName and TestLocales

worldomonation commented on test/e2e/specs/editor-tracking/editor-tracking__block-inserted-event.ts:
```diff
@@ -98,7 +97,7 @@ describe(
 
 					it( '"blog_id" is a numeric value', async function () {
-						expect( numericRegex.test( tracksEventProperties.blog_id ) ).toBe( true );
+						expect( numericRegex.test( tracksEventProperties.blog_id.toString() ) ).toBe( true );
```
It was complaining about values being entered into `expect()` was not a string.

worldomonation commented on packages/calypso-e2e/src/lib/pages/user-signup-page.ts:
```diff
@@ -1,12 +1,7 @@
 import envVariables from '../../env-variables';
-import { NewUserResponse } from '../../rest-api-client';
-
-export interface NewUserDetails {
-	ID: number;
-	bearer_token: string;
-}
```
I ended up removing the `NewUserDetails` interface in favor of returning the whole response instead and let the calling method filter out what they need.

worldomonation committed:
Fix typo

worldomonation committed:
- Move types closer to where it is defined and used.
- Move tests for `lib/utils` out of the `package_root/tests` folder, closer to where the parent modules are.

worldomonation commented:
@dpasque I've revised the PR by moving the type declarations closer to where it is normally consumed. 

I also moved the unit tests under `tests/lib/utils` to `src/lib/utils/tests` directory as part of the change. It's a slight encroachment in the scope of the PR.

dpasque commented on packages/calypso-e2e/src/lib/blocks/block-flows/types.ts:
```diff
@@ -1,5 +1,5 @@
 import { Locator, Page } from 'playwright';
-import { EditorPage } from '../pages';
+import { EditorPage } from '../../pages/editor-page';
```
Small nit --  but we probably want to reference this through the index (`../../pages`) instead of the specific file. That just gives us more flexibility in moving that around in the future with that directory :)

worldomonation replied:
Got it, I'll see if I did this anywhere else and update it wholesale.

dpasque commented on packages/calypso-e2e/src/lib/blocks/block-flows/whatsapp-button.ts:
```diff
@@ -1,4 +1,4 @@
-import { BlockFlow, EditorContext, PublishedPostContext } from '..';
+import { BlockFlow, EditorContext, PublishedPostContext } from './types';
```
Small nit here too -- but you can also let this reference the index here too! E.g. `.`

worldomonation replied:
I've updated this file and all other block-flow files.

dpasque commented on packages/calypso-e2e/src/types/secrets-manager.types.ts:
```diff
@@ -0,0 +1,37 @@
+import { TEST_ACCOUNT_NAMES } from '../secrets';
```
Because the secrets manager has its own directory, can we move this to a file called `types.ts` within `secrets`?

This then will clean up the importing later!

For example, in one place, you have...
```
import { SecretsManager } from '../secrets';
import type { Secrets } from '../types';
```

If this types file was moved into the secrets directory, and added to the index there, all of these could just be imported from `../secrets`!

dpasque commented on packages/calypso-e2e/src/types/env-variables.types.ts:
```diff
@@ -0,0 +1,24 @@
+import { TestAccountName } from '../index';
```
Tiny nit, but I would just reference the directory here, not the specific `index` file name!  `..`

dpasque commented on packages/calypso-e2e/src/lib/utils/types.ts:
```diff
@@ -0,0 +1,5 @@
+// Because these types are ultimately accessed on "window", adding them here.
```
Tiny nit, but we can remove this comment now! :)

worldomonation replied:
Removed.

dpasque commented on test/e2e/specs/editor-tracking/editor-tracking__block-inserted-event.ts:
```diff
@@ -11,15 +11,14 @@ import {
 	HeaderBlock,
 } from '@automattic/calypso-e2e';
 import { Browser, Page } from 'playwright';
+import type { TracksEventProperties } from '@automattic/calypso-e2e';
```
Small nit: but I actually believe you don't have to split out the type implementations here in any of the test scripts, you can just add them to the main import, which should be a little clearer to read 👍

worldomonation replied:
This was/is a preference of mine, since I feel that doing this makes it explicit that this being imported as a type. 

I am primarily going off this response on StackOverflow: https://stackoverflow.com/a/64243357.

dpasque commented on packages/calypso-e2e/src/lib/utils/test/editor-tracks-event-manager.test.ts:
```diff
@@ -1,6 +1,6 @@
 import { it, describe, expect } from '@jest/globals';
```
Required before merge: after moving these tests, they're no longer getting picked up and run in CI.

It looks like you need to update the `jest.config.js` `testMatch` pattern in this package.

I think going from `'<rootDir>/test/**/*.test.ts' ` to `'<rootDir>/src/**/*.test.ts'` should do the trick! 👍

worldomonation replied:
Updated to the suggested pattern.

dpasque reviewed (approved):
Have a few comments!

Required before merge: updating the jest config so the unit tests are included in the CI runs.

The other comments are nits about import syntax and a couple file location changes.

This shouldn't need review again though, so approving! 🚀 

Also, this PR is looking really nice! Thanks for tackling the refactor, and for being open to the suggestion to co-locate the types a bit more 😄

worldomonation committed:
Code review feedback:
- import by index, not by specific file.
- move SecretManager type into `/secrets`

worldomonation committed:
More changes
